Fix: Switch to shadcn form in userResetPassword

AdityaJ2305 opened the pull request:
## Proposed Changes

- Fixes #9683 
- Fixes #9770 
- Fixes #9800 
- Change 1
- Change 2
- More?

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Added a new `PasswordInput` component with password visibility toggle.
	- Enhanced password input handling across multiple forms.
	- Implemented a new password validation mechanism for user password resets.

- **Improvements**
	- Updated password reset and user creation forms with improved validation.
	- Refined error handling for structured API errors.
	- Improved localization messages for password management.
	- Streamlined form submission logic for user password updates.
	- Enhanced flexibility of the `Input` component's onFocus behavior.

- **Bug Fixes**
	- Corrected password validation and submission processes.

- **Chores**
	- Removed deprecated form management components.
	- Removed unused `FormContext` and related types.
	- Removed the `Form` component from the application.
	- Removed the `ExtendPatientRegisterForm` component type from supported plugins.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

AdityaJ2305 committed:
update to shadcn form in userresetpassword

AdityaJ2305 committed:
update to shadcn form in userresetpassword

AdityaJ2305 committed:
deleted old form

AdityaJ2305 committed:
deleted old form

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
fix linting error

AdityaJ2305 committed:
fix linting error

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -13,73 +25,55 @@ import routes from "@/Utils/request/api";
 }) {
   const { t } = useTranslation();
-  const [isSubmitting, setisSubmitting] = useState(false);
+  const [isSubmitting, setIsSubmitting] = useState(false);
```
use `isPending` from `useMutation` instead.

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -13,73 +25,55 @@ import routes from "@/Utils/request/api";
   const { t } = useTranslation();
-  const [isSubmitting, setisSubmitting] = useState(false);
+  const [isSubmitting, setIsSubmitting] = useState(false);
   const [isEditing, setIsEditing] = useState(false);
```
doesn't useForm provide `isDirty` already? this is redundant

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -13,73 +25,55 @@ import routes from "@/Utils/request/api";
 import request from "@/Utils/request/request";
```
switch to `useMutation`

rithviknishad reviewed (changes requested):
you missed one step, one more file was said to be deleted in the issue

AdityaJ2305 committed:
update to mutate

AdityaJ2305 committed:
update to mutate

AdityaJ2305 committed:
update to old_password

AdityaJ2305 committed:
update to old_password

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
ispending

AdityaJ2305 committed:
ispending

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,222 @@
+  });
+  const { mutate: resetUserPasswordMutate, isPending } = useMutation({
+    mutationFn: async (formData: UpdatePasswordForm) => {
+      const response = await mutate(routes.updatePassword, { silent: true })(
+        formData,
+      );
+      if ("errors" in response) {
+        throw response;
+      }
+      return response;
+    },
+    onSuccess: (data: any) => {
+      Notification.Success({ msg: data?.message as string });
+      resetPasswordForm.reset();
+    },
+    onError: (error: any) => {
+      const errorMessage =
+        error?.response?.data?.message || t("password_update_error");
+      Notification.Error({ msg: errorMessage });
+    },
+  });
```
1. follow docs. do not `() => mutate(...)(...)` just do `mutate`
2. `mutate` throws errors for you. you don't need to do it manually.
3. errors notifcations are dispatched by our global error handler. you don't need to do it.


@coderabbitai , maybe update the request's README to include this, since I'm done repeating this same thing on all PRs... :)

AdityaJ2305 committed:
update the error

AdityaJ2305 committed:
update the error

AdityaJ2305 committed:
fix the usemutate

AdityaJ2305 committed:
fix the usemutate

AdityaJ2305 committed:
fix the usemutate

AdityaJ2305 committed:
fix the usemutate

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,214 @@
+    },
+  });
+  const { mutate: resetUserPasswordMutate, isPending } = useMutation({
+    mutationFn: mutate(routes.updatePassword, { silent: true }),
```
why silence it?

AdityaJ2305 replied:
👇

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,214 @@
+      Notification.Success({ msg: data?.message as string });
+      resetPasswordForm.reset();
+    },
+    onError: (error: any) => {
+      const errorMessage =
+        error.cause?.old_password?.[0] ?? t("password_update_error");
+      Notification.Error({ msg: errorMessage });
```
like already said, don't manually dispatch error toasts, it's done automatically

AdityaJ2305 replied:
It is only saying something went wrong not specifying the error like wrong current password

AdityaJ2305 replied:
<img width="452" alt="Screenshot 2025-01-03 at 10 50 52 PM" src="https://github.com/user-attachments/assets/af66c519-7480-4b6b-aed9-4783ae49ccd8" />
<img width="452" alt="Screenshot 2025-01-03 at 10 50 08 PM" src="https://github.com/user-attachments/assets/faeaad72-7bd1-47db-bb30-d54e3410a9ec" />

rithviknishad replied:
make the global error handler do it. let's not write custom logics.

AdityaJ2305 replied:
But it is specific for this comp for wrong password error in line no. 83

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,214 @@
+  const { mutate: resetUserPasswordMutate, isPending } = useMutation({
+    mutationFn: mutate(routes.updatePassword, { silent: true }),
+    onSuccess: (data: any) => {
+      Notification.Success({ msg: data?.message as string });
```
use toast.success instead. these are deprecated functions.

rithviknishad reviewed (changes requested):
repeating the second time, do delete the other file that was mentioned in the issue

AdityaJ2305 commented:
> repeating the second time, do delete the other file that was mentioned in the issue

the context file is used by other file like `pluginTypes.ts`

rithviknishad commented:
delete those usages 🙂 we don't need it. let's get rid of weird things.

AdityaJ2305 committed:
use of toast and update to use translation

AdityaJ2305 committed:
use of toast and update to use translation

AdityaJ2305 commented:
> delete those usages 🙂 we don't need it. let's get rid of weird things.

https://github.com/ohcnetwork/care_fe/blob/b36f39fea38d18425dcad74a8d43b29ae8eee1a7/src/pluginTypes.ts#L42

rithviknishad commented:
yup, you can remove that patient register form extension thing.... 

https://github.com/ohcnetwork/care_fe/blob/b36f39fea38d18425dcad74a8d43b29ae8eee1a7/src%2FpluginTypes.ts#L30-L43

AdityaJ2305 committed:
deleted formcontext

AdityaJ2305 committed:
deleted formcontext

AdityaJ2305 commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,222 @@
+    },
+    onError: (error: any) => {
+      const errorMessage =
+        error.cause?.old_password?.[0] ?? t("password_update_error");
```
@rithviknishad

AdityaJ2305 committed:
update global error handler

AdityaJ2305 committed:
update global error handler

AdityaJ2305 committed:
update global error handler

AdityaJ2305 committed:
update global error handler

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/Utils/request/errorHandler.ts:
why is there so many lines changed in this file? ensure things relevant are only changed

AdityaJ2305 replied:
just because of prettier

rithviknishad replied:
why is the functions moved inside the function?

AdityaJ2305 committed:
fix bracket

AdityaJ2305 committed:
fix bracket

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -32,11 +32,19 @@ export function handleHttpError(error: Error) {
   if (isBadRequest(error)) {
     const errs = cause?.errors;
-    if (isPydanticError(errs)) {
+    if (errs && isPydanticError(errs)) {
```
isPydanticError will anyways return false if errs is false right?

```suggestion
    if (isPydanticError(errs)) {
```

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -32,11 +32,19 @@ export function handleHttpError(error: Error) {
     }
-    Notifications.BadRequest({ errs });
+    if (cause && typeof cause === "object" && !Array.isArray(cause)) {
+      const firstKey = Object.keys(cause)[0];
+      const firstError = firstKey ? cause[firstKey] : null;
+      if (firstError && Array.isArray(firstError)) {
+        toast.error(firstError[0]);
+        return;
+      }
```
why limit to just first error? what if backend gives more errors?

also move this to a separate function like other checks and handlers. let's not make the global error handler messy by inlining things. not readable code.

AdityaJ2305 committed:
fix bracket

AdityaJ2305 committed:
fix bracket

AdityaJ2305 committed:
Update src/Utils/request/errorHandler.ts

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

AdityaJ2305 committed:
Update src/Utils/request/errorHandler.ts

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

AdityaJ2305 committed:
move to seperate func

AdityaJ2305 committed:
move to seperate func

AdityaJ2305 committed:
lint fixed

AdityaJ2305 committed:
lint fixed

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AnveshNalimela committed:
Added no user assingned message (#9666)

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

Mahendar0701 committed:
Remove facility name from facility users page (#9746)

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,217 @@
-    setisSubmitting(true);
+  const [isPasswordFieldFocused, setIsPasswordFieldFocused] = useState(false);
+
+  const resetPasswordForm = useForm({
```
the file is anyways named `UserResetPassword.tsx`, let's keep the names simple since there's no other forms in the same file.

```suggestion
  const form = useForm({
```

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,217 @@
+  });
+  const { mutate: resetUserPasswordMutate, isPending } = useMutation({
+    mutationFn: mutate(routes.updatePassword),
+    onSuccess: (data: any) => {
+      toast.success(data?.message as string);
+      resetPasswordForm.reset();
+    },
```
don't use `any` type. use types infered by the routes schema
```suggestion
    onSuccess: (data) => {
      toast.success(data?.message as string);
      resetPasswordForm.reset();
    },
```

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,217 @@
+      new_password_2: "",
+    },
+  });
+  const { mutate: resetUserPasswordMutate, isPending } = useMutation({
```
```suggestion
  const { mutate: resetPassword, isPending } = useMutation({
```

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,207 +1,217 @@
-                  field("new_password_1").value?.length >= 8,
-                  t("password_length_validation"),
-                  !field("new_password_1").value,
+  const renderPasswordForm = () => (
```
why not just inline this content directly inside where this is called?

rithviknishad reviewed (commented):
mostly lgtm, some minor changes

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -97,3 +92,25 @@ function handlePydanticErrors(errors: PydanticError[]) {
   });
 }
+
+function handleBadRequest(cause: any) {
```
do not use `any` type please....

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -97,3 +92,25 @@ function handlePydanticErrors(errors: PydanticError[]) {
+    return;
+  }
+
+  if (cause && typeof cause === "object" && !Array.isArray(cause)) {
+    Object.values(cause).forEach((value) => {
+      if (Array.isArray(value)) {
+        value.forEach((err) => {
+          toast.error(`${err}`);
+        });
+      }
+    });
+    return;
+  }
```
what i meant was, write checks and handler for this piece of logic alone, and not to move the entire thing as handleBadRequest. Code ain't readable this way.

bodhish committed:
Enhance organization management by adding new organization types and updating related components. Refactor organization level retrieval to use metadata for improved localization. Remove deprecated organization levels constant and clean up unused code in various components.

arhamathar committed:
Cancel button fix in Book Appointment screen (#9757)

bodhish committed:
Add public flag to facility (#9759)

abhimanyurajeesh committed:
Auto-Hide Sidebar on mobile (#9758)

AdityaJ2305 committed:
fix the types and typo changes and strucutured code

AdityaJ2305 committed:
fix the types and typo changes and strucutured code

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

bodhish committed:
Implement Permission Context and Update User Permissions Handling

- Introduced a new PermissionContext to manage user permissions and super admin status across the application.
- Updated AppRouter to utilize PermissionProvider for managing permissions.
- Changed UserModel to store permissions as strings instead of objects.
- Refactored OrganizationFacilities and OrganizationPatients components to use updated permission checks and organization identifiers.
- Enhanced OrganizationLayout to conditionally render navigation items based on user permissions.

This commit improves the permission management system and ensures consistent handling of user permissions throughout the application.

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -76,6 +82,24 @@ type PydanticError = {
+function handleStructuredErrors(cause: HTTPErrorCause) {
+  if (cause && typeof cause === "object" && !Array.isArray(cause)) {
+    Object.values(cause).forEach((value) => {
+      if (Array.isArray(value)) {
+        value.forEach((err) => {
+          toast.error(`${err}`);
+        });
+      }
```
and what if it's not an array, but a string? that also needs to be handled right?

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -76,6 +82,24 @@ type PydanticError = {
+  );
+}
+
+function handleStructuredErrors(cause: HTTPErrorCause) {
```
```suggestion
function handleStructuredErrors(cause: HttpError["cause"]) {
```

rithviknishad commented on src/Utils/request/types.ts:
```diff
@@ -51,7 +51,7 @@ export interface APICallOptions<TBody = unknown> {
 }
 
-type HTTPErrorCause = Record<string, unknown> | undefined;
+export type HTTPErrorCause = Record<string, unknown> | undefined;
```
```suggestion
type HTTPErrorCause = Record<string, unknown> | undefined;
```

AdityaJ2305 committed:
fix the string error hadler

AdityaJ2305 committed:
fix the string error hadler

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -76,6 +82,24 @@ type PydanticError = {
   url: string;
 };
 
+function isStructuredError(cause: HTTPError["cause"]): boolean {
+  return typeof cause === "object" && !Array.isArray(cause);
+}
+
+function handleStructuredErrors(cause: HTTPError["cause"]) {
+  if (cause && typeof cause === "object" && !Array.isArray(cause)) {
+    Object.values(cause).forEach((value) => {
+      if (Array.isArray(value)) {
+        value.forEach((err) => {
+          toast.error(`${err}`);
+        });
+      } else if (typeof value === "string") {
+        toast.error(value);
+      }
+    });
+  }
+}
```
1. this check is already done by `isStructuredError`. why check again inside the handler?
you can make the `cause` type of handler more restrictive to `Record<string, string | string[]>`
and update the return type of isStructeredError to `cause is Record<string, string | string[]>` instead.
similar to how pydantic checks and handlers are done:

<img width="878" alt="image" src="https://github.com/user-attachments/assets/c02970d2-fab6-46ae-ad71-db0ce7aad281" />

2. Also prefer using for loop so that we can make use of early returns to avoid if else ladders, and we can extend for more structure later on if needed.

```suggestion
function isStructuredError(cause: HTTPError["cause"]): cause is Record<string, string | string[]> {
  return typeof cause === "object" && !Array.isArray(cause);
}

function handleStructuredErrors(cause: Record<string, string | string[]>) {
  for (const value of Object.values(cause)) {
    if (Array.isArray(value)) {
      value.forEach((err) => toast.error(err));
      return;
    }
    if (typeof value === "string") {
      toast.error(value);
      return;
    }
  }
}
```

AdityaJ2305 committed:
fix the structure of handlers

AdityaJ2305 committed:
fix the structure of handlers

AdityaJ2305 committed:
Empty-Commit

AdityaJ2305 committed:
Empty-Commit

AdityaJ2305 committed:
fix merge conflicts

AdityaJ2305 committed:
fix merge conflicts

AdityaJ2305 committed:
fix merge conflicts

AdityaJ2305 committed:
fix merge conflicts

rithviknishad committed:
keep things simpler

rithviknishad committed:
keep things simpler

rithviknishad commented:
do clear conflicts

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
fix merge conflicts

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
fix lint sfter merge

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
fix lint sfter merge

AdityaJ2305 commented:
> do clear conflicts

@rithviknishad done

rithviknishad commented:
it doesn't inform the user what the error is when i click submit.

do a self test of all the cases after making the changes and then request for review.

Also let's keep it as grid with two columns (in desktop view) itself like how it was before. New and Confirm password can be in second row though however.

<img width="1289" alt="image" src="https://github.com/user-attachments/assets/5afe1e83-bc34-413a-a68b-15e57f2e0606" />

rithviknishad commented on src/components/Users/UserResetPassword.tsx:
```diff
@@ -1,216 +1,219 @@
+                  <FormItem>
+                    <FormLabel>{t("old_password")}</FormLabel>
+                    <FormControl>
+                      <TextFormField
+                        {...field}
+                        type="password"
+                        onChange={(value) => {
+                          field.onChange(value.value);
+                        }}
+                      />
```
also use shadcn's input component instead. FormField components are not intended to be used along with shadcn's form component.

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
shift to shadcn input and add the eye toggle

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
same password validation check

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/components/ui/input.tsx:
do not modify shadcn's input component. you may extend it by creating a new component instead.

rithviknishad replied:
also why was the login page not updated like mentioned in the issue?

also update in the user create form

AdityaJ2305 replied:
@rithviknishad, understood. The eye toggle functionality seems to be working on both the login page and the user creation form for the password field. Am I misunderstanding the issue? 😅 However, it does cause some UI misalignment for components like +91 in the Patient Login, which could be addressed once the separator component is created.

Here are the ss of deploy preview 

<img width="1319" alt="Screenshot 2025-01-06 at 1 30 56 PM" src="https://github.com/user-attachments/assets/0e4ecb61-bf6c-4fa2-8842-577c7c6ecfa9" />
<img width="1319" alt="Screenshot 2025-01-06 at 1 34 27 PM" src="https://github.com/user-attachments/assets/18dc2f28-95b6-4134-b754-b1bf23e03cc8" />

rithviknishad replied:
oh my bad. probably a wrong deploy preview was used.

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
reset input field

AdityaJ2305 committed:
create passport input

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
create passport input

AdityaJ2305 committed:
update to passwordinput

AdityaJ2305 committed:
update ResetPassword.tsx

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
update ResetPassword.tsx

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
lint error

AdityaJ2305 committed:
update the design and fix zod issue

AdityaJ2305 committed:
update the design and fix zod issue

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/components/ui/input-password.tsx:
```diff
@@ -0,0 +1,37 @@
+        className={cn("pr-10", className)}
+        ref={ref}
+        {...props}
+        onFocus={props.onFocus}
+        onBlur={props.onBlur}
```
the props spread just before this already does it, why pass it again?

```suggestion
```

AdityaJ2305 replied:
Done

rithviknishad commented on src/components/ui/input.tsx:
like already said, do not modify shadcn's component

AdityaJ2305 replied:
@rithviknishad, we can remove the `onBlur` from the `ShadCN input` as spread operator doing that for us , but since `onFocus`  isn’t set up to accept a custom `onFocus` function from props as it is overriding the `onFocus` from props. we’d need to modify.

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
remove unnecessary code

rithviknishad commented on src/components/ui/input.tsx:
```diff
@@ -17,6 +17,9 @@ const Input = React.forwardRef<HTMLInputElement, React.ComponentProps<"input">>(
           if (type === "date" || type === "time") {
             e.target.showPicker();
           }
+          if (props.onFocus) {
+            props.onFocus(e);
+          }
```
```suggestion
          props.onFocus?.(e);
```

rithviknishad committed:
Update src/components/ui/input.tsx

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

Jacobjeevan commented on src/Utils/request/errorHandler.ts:
```diff
@@ -36,6 +36,12 @@ export function handleHttpError(error: Error) {
       handlePydanticErrors(errs);
```
You will need to modify handlePydanticErrors to get this working; it will fail when loc is undefined.

```
const location = loc?.join(".") ? `: '${loc.join(".")}'` : "";
const message = typeof msg === "object" ? (Object.values(msg)[0] as string) : msg;
toast.error(`${title}${location}`, {
      description: message,
      duration: 8000,
    });
```

<img width="406" alt="image" src="https://github.com/user-attachments/assets/67e63a46-76bb-468d-a40d-2aa0e43c37f5" />

We might need to modify the handleHttpError to be a functional component (to add support for translations?).
Though that might be unnecessary given we would have FE validations anyways.

cc: @rithviknishad

rithviknishad replied:
The loc undefined issue has been solved here:

https://github.com/ohcnetwork/care_fe/pull/9799/files#diff-fdac5dcec1799cbf5217e79c96d50860c2225a9449e5fe976844e33780501d01

rithviknishad replied:
@AdityaJ2305 maybe cherrypick that file changes from my PR alone here.

Jacobjeevan replied:
Following up, just talked to Rithvik as well  - msg can be an object, so you need to handle accordingly, so get the first value (there should only be one anyways).

To clarify, you will have cases where loc is undefined and msg is an object (off the top of my head, I don't recall if there are cases in the new BE where msg is a string, but let's handle that to be safe).

rithviknishad replied:
yup, there are cases where msg is string too, when non field errors are there. I made that change coz scheduling had it.

AdityaJ2305 replied:
@Jacobjeevan @rithviknishad, I’ve applied the same changes from Rithvik’s draft PR.

AdityaJ2305 replied:
@Jacobjeevan, just a quick doubt—based on the screenshot you shared, I’m assuming you’re on the `UserCreateForm` . Since it already has frontend validations, did you try removing them to test? Or is it from a different form?

AdityaJ2305 replied:
<img width="645" alt="Screenshot 2025-01-07 at 4 00 48 PM" src="https://github.com/user-attachments/assets/75258310-ae1f-449d-a8de-1b34478d07e4" />

Jacobjeevan replied:
> @Jacobjeevan, just a quick doubt—based on the screenshot you shared, I’m assuming you’re on the `UserCreateForm` . Since it already has frontend validations, did you try removing them to test? Or is it from a different form?

yep

Jacobjeevan replied:
> Following up, just talked to Rithvik as well - msg can be an object, so you need to handle accordingly, so get the first value (there should only be one anyways).
> 
> To clarify, you will have cases where loc is undefined and msg is an object (off the top of my head, I don't recall if there are cases in the new BE where msg is a string, but let's handle that to be safe).

☝️

Jacobjeevan replied:
You need to handle both

(1) Loc is undefined
(2) Msg is object or string.

AdityaJ2305 committed:
added code changes from rithvik pr

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
handle msg object string

AdityaJ2305 committed:
Merge branch 'issue/9683/userResetPassword_form_switch' of https://github.com/AdityaJ2305/care_fe into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/Utils/request/errorHandler.ts:
```diff
@@ -87,12 +110,16 @@ function isPydanticError(errors: unknown): errors is PydanticError[] {
 function handlePydanticErrors(errors: PydanticError[]) {
   errors.map(({ type, loc, msg }) => {
-    const title = type
+    const message = typeof msg === "string" ? msg : Object.values(msg)[0];
+    if (!loc) {
+      toast.error(message);
+      return;
+    }
+    type = type
```
not a huge fan of this, but conditionally approving for now. we need to improve how we handle errors later on, there are couple of edge cases missing here, and cases where that's not supposed to be handled here too.

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

nihal467 commented:
Approving it conditionally, as we don't have email-based reset password page

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
resolved conflicts

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
valid text added

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
updated the code

AdityaJ2305 committed:
updated the code

AdityaJ2305 commented:
TODO: Need to update the messages for `CreateUserForm` similar to  `Use at least 8 characters for your password.`

AdityaJ2305 committed:
resolved conflicts

AdityaJ2305 committed:
resolved conflicts

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
resolved conflicts

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

rithviknishad commented on src/components/Users/CreateUserForm.tsx:
```diff
@@ -38,70 +39,64 @@ import { UserBase } from "@/types/user/user";
+        ),
+      password: z
+        .string()
+        .min(8, "Password must be at least 8 characters")
+        .regex(/[a-z]/, "Password must contain at least one lowercase letter")
+        .regex(/[A-Z]/, "Password must contain at least one uppercase letter")
+        .regex(/[0-9]/, "Password must contain at least one number"),
+      c_password: z.string(),
+      first_name: z.string().min(1, "First name is required"),
+      last_name: z.string().min(1, "Last name is required"),
+      email: z.string().email("Invalid email address"),
+      phone_number: z
+        .string()
+        .regex(
+          /^\+91[0-9]{10}$/,
+          "Phone number must start with +91 followed by 10 digits",
+        ),
```
missing i18n

rithviknishad commented on src/components/Users/CreateUserForm.tsx:
```diff
@@ -127,7 +122,7 @@ export default function CreateUserForm({ onSubmitSuccess }: Props) {
   }, [phoneNumber, isWhatsApp, form, usernameInput]);
 
-  const { error, isLoading } = useQuery({
+  const { isLoading, isError } = useQuery({
```
since this is not the primary API of this component, either alias the isLoading, isError, or skip deconstructing it.
```suggestion
  const usernameCheckQuery = useQuery({
```

AdityaJ2305 replied:
Let's go with alias

rithviknishad commented on src/components/ui/input-password.tsx:
```diff
@@ -0,0 +1,35 @@
+
+import CareIcon from "@/CAREUI/icons/CareIcon";
+
+import { Input } from "./input";
```
absolute imports

rithviknishad reviewed (approved):
lgtm, some minor things. let's get it merged after that

AdityaJ2305 committed:
Merge branch 'develop' into issue/9683/userResetPassword_form_switch

AdityaJ2305 committed:
done minor changes

AdityaJ2305 commented:
@rithviknishad done with the minor changes

rithviknishad commented:
for required fields, we reuse "field_required" instead of creating very specific translations for each field

AdityaJ2305 committed:
this field is req
